src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	// local copy of their type handles first (else, as we log them we could flush
+	// and clear out m_rg_bulk_type_values, thus trashing p_val)
+	uint32_t c_params = p_val->c_type_parameters;
+	intptr_t rg_type_parameters[c_params];
+	MonoType *rg_mono_type_parameters[c_params];
+	for (uint32_t i = 0; i < c_params; i++)
+	{
+		rg_type_parameters[i] = p_val->rg_type_parameters[i];
+		rg_mono_type_parameters[i] = p_val->rg_mono_type_parameters[i];
+	}
+
+	for (uint32_t i = 0; i < c_params; i++)
+		ep_rt_mono_log_type_and_parameters_if_necessary (p_type_logger, rg_mono_type_parameters[i], rg_type_parameters[i]);
+}
+
+//---------------------------------------------------------------------------------------
+//
+// Outermost level of ETW-type-logging.  This method is used to log a unique type identifier
+// (in this case a MonoType) and (recursively) its type parameters when present.
+//
+// Arguments:
+//      * mono_type - MonoType to be logged
+//      * type_id - Unique type identifier of mono_type
+
+// static
+void
+ep_rt_mono_log_type_and_parameters_if_necessary (BulkTypeEventLogger *p_type_logger, MonoType *mono_type, intptr_t type_id)
+{
+	// TODO Log the type if necessary
+
+	ep_rt_mono_log_type_and_parameters (p_type_logger, mono_type, type_id);
+}
+
+//---------------------------------------------------------------------------------------
+//
+// ep_rt_mono_send_method_details_event is the method responsible for sending details of
+// methods involved in events such as JitStart, Load/Unload, Rundown, R2R, and other
+// eventpipe events. It calls ep_rt_mono_log_type_and_parameters_if_necessary to log
+// unique types from the method type and available method instantiation parameter types
+// that are ultimately emitted as a BulkType event in ep_rt_mono_fire_bulk_type_event.
+// After appropraitely logging type information, it sends method details outlined by
+// the generated dotnetruntime.c and ClrEtwAll manifest.
+//
+// Arguments:
+//      * method - a MonoMethod hit during an eventpipe event
+
+void
+ep_rt_mono_send_method_details_event (MonoMethod *method)
+{
+	if (method->dynamic)
+		return;
+
+	MonoGenericContext *method_ctx = mono_method_get_context(method);
+
+	MonoGenericInst *method_inst = NULL;
+	if (method_ctx)
+		method_inst = method_ctx->method_inst;
+
+	if (method_inst)
+		if (method_inst->type_argc > 1024) // ETW has a limit for maximum event size. Do not log overly large method type argument sets
+			return;
+
+	BulkTypeEventLogger *p_type_logger = malloc(sizeof(BulkTypeEventLogger));
+
+	intptr_t method_type_id = 0;
+	uint32_t method_token = (method->token & 0xFFFFFF) | 0x06000000; // dotnet-pgo ResolveMethodID expects method tokens with a 0x06 mask
+	uint64_t loader_module_id = 0;
+	MonoClass *klass = method->klass;
+	if (klass)
+	{
+		MonoType *method_mono_type = m_class_get_byval_arg (klass);
+		method_type_id = get_typeid_for_type (method_mono_type);
```

Comment:
This pattern seems pretty common.  Maybe add `get_type_id_for_class` as a helper?

Reply:
Added! I think it applies twice.